### PR TITLE
Fix bug #182 printf %T  "ago" broken

### PR DESCRIPTION
**Pull request by phidebian**
This is the PR for #182

### CODE REVIEW

**Comment by McDutchie** [comment]
Thanks for the PR! I'll test it now. Just a few things:
* Can you explain your changes? What does each change do and what is the reason for it? This should be documented in the git commit log, so that others (including myself) can understand what's going on later and possibly build on your fixes. (For now you can just add a comment here, I'll make sure it gets added to the commit log.)
* Please use tabs for indentation, not spaces.
* Please use the same coding style as the surrounding code.

**Comment by McDutchie** [comment]
I've tested your changes with my reproducer from [this comment](https://github.com/ksh93/ksh/issues/182#issue-810912212). Before the PR, 268 iterations failed. After the PR, 209 still fail. Here is an output of the reproducer with the PR.

```
[BUG]  24   years ago: ksh: 1998-01-01 00:00:00 != gnu: 1998-10-18 03:29:36
[BUG]  23   years ago: ksh: 1999-01-01 00:00:00 != gnu: 1999-10-18 03:29:36
[BUG]  22   years ago: ksh: 2000-01-01 00:00:00 != gnu: 2000-10-18 03:29:36
[BUG]  21   years ago: ksh: 2001-01-01 00:00:00 != gnu: 2001-10-18 03:29:36
[BUG]  20   years ago: ksh: 2002-01-01 00:00:00 != gnu: 2002-10-18 03:29:36
[BUG]  19   years ago: ksh: 2003-01-01 00:00:00 != gnu: 2003-10-18 03:29:36
[BUG]  18   years ago: ksh: 2004-01-01 00:00:00 != gnu: 2004-10-18 03:29:36
[BUG]  17   years ago: ksh: 2005-01-01 00:00:00 != gnu: 2005-10-18 03:29:36
[BUG]  16   years ago: ksh: 2006-01-01 00:00:00 != gnu: 2006-10-18 03:29:36
[BUG]  15   years ago: ksh: 2007-01-01 00:00:00 != gnu: 2007-10-18 03:29:36
[BUG]  14   years ago: ksh: 2008-01-01 00:00:00 != gnu: 2008-10-18 03:29:36
[BUG]  13   years ago: ksh: 2009-01-01 00:00:00 != gnu: 2009-10-18 03:29:36
[BUG]  12   years ago: ksh: 2010-01-01 00:00:00 != gnu: 2010-10-18 03:29:36
[BUG]  11   years ago: ksh: 2011-01-01 00:00:00 != gnu: 2011-10-18 03:29:36
[BUG]  10   years ago: ksh: 2012-01-01 00:00:00 != gnu: 2012-10-18 03:29:36
[BUG]   9   years ago: ksh: 2013-01-01 00:00:00 != gnu: 2013-10-18 03:29:36
[BUG]   8   years ago: ksh: 2014-01-01 00:00:00 != gnu: 2014-10-18 03:29:36
[BUG]   7   years ago: ksh: 2015-01-01 00:00:00 != gnu: 2015-10-18 03:29:36
[BUG]   6   years ago: ksh: 2016-01-01 00:00:00 != gnu: 2016-10-18 03:29:36
[BUG]   5   years ago: ksh: 2017-01-01 00:00:00 != gnu: 2017-10-18 03:29:36
[BUG]   4   years ago: ksh: 2018-01-01 00:00:00 != gnu: 2018-10-18 03:29:36
[BUG]   3   years ago: ksh: 2019-01-01 00:00:00 != gnu: 2019-10-18 03:29:36
[BUG]   2   years ago: ksh: 2020-01-01 00:00:00 != gnu: 2020-10-18 03:29:36
[BUG]   1   years ago: ksh: 2021-01-01 00:00:00 != gnu: 2021-10-18 03:29:36
[BUG]   0   years ago: ksh: 2022-01-01 00:00:00 != gnu: 2022-10-18 03:29:36
[BUG]  -1   years ago: ksh: 2023-01-01 00:00:00 != gnu: 2023-10-18 03:29:36
[BUG]  -2   years ago: ksh: 2024-01-01 00:00:00 != gnu: 2024-10-18 03:29:36
[BUG]  -3   years ago: ksh: 2025-01-01 00:00:00 != gnu: 2025-10-18 03:29:36
[BUG]  -4   years ago: ksh: 2026-01-01 00:00:00 != gnu: 2026-10-18 03:29:36
[BUG]  -5   years ago: ksh: 2027-01-01 00:00:00 != gnu: 2027-10-18 03:29:36
[BUG]  -6   years ago: ksh: 2028-01-01 00:00:00 != gnu: 2028-10-18 03:29:36
[BUG]  -7   years ago: ksh: 2029-01-01 00:00:00 != gnu: 2029-10-18 03:29:36
[BUG]  -8   years ago: ksh: 2030-01-01 00:00:00 != gnu: 2030-10-18 03:29:36
[BUG]  -9   years ago: ksh: 2031-01-01 00:00:00 != gnu: 2031-10-18 03:29:36
[BUG] -10   years ago: ksh: 2032-01-01 00:00:00 != gnu: 2032-10-18 03:29:36
[BUG] -11   years ago: ksh: 2033-01-01 00:00:00 != gnu: 2033-10-18 03:29:36
[BUG] -12   years ago: ksh: 2034-01-01 00:00:00 != gnu: 2034-10-18 03:29:36
[BUG] -13   years ago: ksh: 2035-01-01 00:00:00 != gnu: 2035-10-18 03:29:36
[BUG] -14   years ago: ksh: 2036-01-01 00:00:00 != gnu: 2036-10-18 03:29:36
[BUG] -15   years ago: ksh: 2037-01-01 00:00:00 != gnu: 2037-10-18 03:29:36
[BUG] -16   years ago: ksh: 2038-01-01 00:00:00 != gnu: 2038-10-18 03:29:36
[BUG] -17   years ago: ksh: 2039-01-01 00:00:00 != gnu: 2039-10-18 03:29:36
[BUG] -18   years ago: ksh: 2040-01-01 00:00:00 != gnu: 2040-10-18 03:29:36
[BUG] -19   years ago: ksh: 2041-01-01 00:00:00 != gnu: 2041-10-18 03:29:36
[BUG] -20   years ago: ksh: 2042-01-01 00:00:00 != gnu: 2042-10-18 03:29:36
[BUG] -21   years ago: ksh: 2043-01-01 00:00:00 != gnu: 2043-10-18 03:29:36
[BUG] -22   years ago: ksh: 2044-01-01 00:00:00 != gnu: 2044-10-18 03:29:36
[BUG] -23   years ago: ksh: 2045-01-01 00:00:00 != gnu: 2045-10-18 03:29:36
[BUG] -24   years ago: ksh: 2046-01-01 00:00:00 != gnu: 2046-10-18 03:29:36
[BUG]  24  months ago: ksh: 2020-10-01 00:00:00 != gnu: 2020-10-18 03:29:36
[BUG]  23  months ago: ksh: 2020-11-01 00:00:00 != gnu: 2020-11-18 02:29:36
[BUG]  22  months ago: ksh: 2020-12-01 00:00:00 != gnu: 2020-12-18 02:29:36
[BUG]  21  months ago: ksh: 2021-01-01 00:00:00 != gnu: 2021-01-18 02:29:36
[BUG]  20  months ago: ksh: 2021-02-01 00:00:00 != gnu: 2021-02-18 02:29:36
[BUG]  19  months ago: ksh: 2021-03-01 00:00:00 != gnu: 2021-03-18 02:29:36
[BUG]  18  months ago: ksh: 2021-04-01 00:00:00 != gnu: 2021-04-18 03:29:36
[BUG]  17  months ago: ksh: 2021-05-01 00:00:00 != gnu: 2021-05-18 03:29:36
[BUG]  16  months ago: ksh: 2021-06-01 00:00:00 != gnu: 2021-06-18 03:29:36
[BUG]  15  months ago: ksh: 2021-07-01 00:00:00 != gnu: 2021-07-18 03:29:36
[BUG]  14  months ago: ksh: 2021-08-01 00:00:00 != gnu: 2021-08-18 03:29:36
[BUG]  13  months ago: ksh: 2021-09-01 00:00:00 != gnu: 2021-09-18 03:29:36
[BUG]  12  months ago: ksh: 2021-10-01 00:00:00 != gnu: 2021-10-18 03:29:36
[BUG]  11  months ago: ksh: 2021-11-01 00:00:00 != gnu: 2021-11-18 02:29:36
[BUG]  10  months ago: ksh: 2021-12-01 00:00:00 != gnu: 2021-12-18 02:29:36
[BUG]   9  months ago: ksh: 2022-01-01 00:00:00 != gnu: 2022-01-18 02:29:36
[BUG]   8  months ago: ksh: 2022-02-01 00:00:00 != gnu: 2022-02-18 02:29:36
[BUG]   7  months ago: ksh: 2022-03-01 00:00:00 != gnu: 2022-03-18 02:29:36
[BUG]   6  months ago: ksh: 2022-04-01 00:00:00 != gnu: 2022-04-18 03:29:36
[BUG]   5  months ago: ksh: 2022-05-01 00:00:00 != gnu: 2022-05-18 03:29:36
[BUG]   4  months ago: ksh: 2022-06-01 00:00:00 != gnu: 2022-06-18 03:29:36
[BUG]   3  months ago: ksh: 2022-07-01 00:00:00 != gnu: 2022-07-18 03:29:36
[BUG]   2  months ago: ksh: 2022-08-01 00:00:00 != gnu: 2022-08-18 03:29:36
[BUG]   1  months ago: ksh: 2022-09-01 00:00:00 != gnu: 2022-09-18 03:29:36
[BUG]   0  months ago: ksh: 2022-10-01 00:00:00 != gnu: 2022-10-18 03:29:36
[BUG]  -1  months ago: ksh: 2022-11-01 00:00:00 != gnu: 2022-11-18 02:29:36
[BUG]  -2  months ago: ksh: 2022-12-01 00:00:00 != gnu: 2022-12-18 02:29:36
[BUG]  -3  months ago: ksh: 2023-01-01 00:00:00 != gnu: 2023-01-18 02:29:36
[BUG]  -4  months ago: ksh: 2023-02-01 00:00:00 != gnu: 2023-02-18 02:29:36
[BUG]  -5  months ago: ksh: 2023-03-01 00:00:00 != gnu: 2023-03-18 02:29:36
[BUG]  -6  months ago: ksh: 2023-04-01 00:00:00 != gnu: 2023-04-18 03:29:36
[BUG]  -7  months ago: ksh: 2023-05-01 00:00:00 != gnu: 2023-05-18 03:29:36
[BUG]  -8  months ago: ksh: 2023-06-01 00:00:00 != gnu: 2023-06-18 03:29:36
[BUG]  -9  months ago: ksh: 2023-07-01 00:00:00 != gnu: 2023-07-18 03:29:36
[BUG] -10  months ago: ksh: 2023-08-01 00:00:00 != gnu: 2023-08-18 03:29:36
[BUG] -11  months ago: ksh: 2023-09-01 00:00:00 != gnu: 2023-09-18 03:29:36
[BUG] -12  months ago: ksh: 2023-10-01 00:00:00 != gnu: 2023-10-18 03:29:36
[BUG] -13  months ago: ksh: 2023-11-01 00:00:00 != gnu: 2023-11-18 02:29:36
[BUG] -14  months ago: ksh: 2023-12-01 00:00:00 != gnu: 2023-12-18 02:29:36
[BUG] -15  months ago: ksh: 2024-01-01 00:00:00 != gnu: 2024-01-18 02:29:36
[BUG] -16  months ago: ksh: 2024-02-01 00:00:00 != gnu: 2024-02-18 02:29:36
[BUG] -17  months ago: ksh: 2024-03-01 00:00:00 != gnu: 2024-03-18 02:29:36
[BUG] -18  months ago: ksh: 2024-04-01 00:00:00 != gnu: 2024-04-18 03:29:36
[BUG] -19  months ago: ksh: 2024-05-01 00:00:00 != gnu: 2024-05-18 03:29:36
[BUG] -20  months ago: ksh: 2024-06-01 00:00:00 != gnu: 2024-06-18 03:29:36
[BUG] -21  months ago: ksh: 2024-07-01 00:00:00 != gnu: 2024-07-18 03:29:36
[BUG] -22  months ago: ksh: 2024-08-01 00:00:00 != gnu: 2024-08-18 03:29:36
[BUG] -23  months ago: ksh: 2024-09-01 00:00:00 != gnu: 2024-09-18 03:29:36
[BUG] -24  months ago: ksh: 2024-10-01 00:00:00 != gnu: 2024-10-18 03:29:36
[BUG]  24   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-05-03 03:29:36
[BUG]  23   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-05-10 03:29:36
[BUG]  22   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-05-17 03:29:36
[BUG]  21   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-05-24 03:29:36
[BUG]  20   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-05-31 03:29:36
[BUG]  19   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-06-07 03:29:36
[BUG]  18   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-06-14 03:29:36
[BUG]  17   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-06-21 03:29:36
[BUG]  16   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-06-28 03:29:36
[BUG]  15   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-07-05 03:29:36
[BUG]  14   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-07-12 03:29:36
[BUG]  13   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-07-19 03:29:36
[BUG]  12   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-07-26 03:29:36
[BUG]  11   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-08-02 03:29:36
[BUG]  10   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-08-09 03:29:36
[BUG]   9   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-08-16 03:29:36
[BUG]   8   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-08-23 03:29:36
[BUG]   7   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-08-30 03:29:36
[BUG]   6   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-09-06 03:29:36
[BUG]   5   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-09-13 03:29:36
[BUG]   4   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-09-20 03:29:36
[BUG]   3   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-09-27 03:29:36
[BUG]   2   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-10-04 03:29:36
[BUG]   1   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-10-11 03:29:36
[BUG]   0   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-10-18 03:29:36
[BUG]  -1   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-10-25 03:29:36
[BUG]  -2   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-11-01 02:29:36
[BUG]  -3   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-11-08 02:29:36
[BUG]  -4   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-11-15 02:29:36
[BUG]  -5   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-11-22 02:29:36
[BUG]  -6   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-11-29 02:29:36
[BUG]  -7   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-12-06 02:29:36
[BUG]  -8   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-12-13 02:29:36
[BUG]  -9   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-12-20 02:29:36
[BUG] -10   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2022-12-27 02:29:36
[BUG] -11   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-01-03 02:29:36
[BUG] -12   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-01-10 02:29:36
[BUG] -13   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-01-17 02:29:36
[BUG] -14   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-01-24 02:29:36
[BUG] -15   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-01-31 02:29:36
[BUG] -16   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-02-07 02:29:36
[BUG] -17   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-02-14 02:29:36
[BUG] -18   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-02-21 02:29:36
[BUG] -19   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-02-28 02:29:36
[BUG] -20   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-03-07 02:29:37
[BUG] -21   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-03-14 02:29:37
[BUG] -22   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-03-21 02:29:37
[BUG] -23   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-03-28 03:29:37
[BUG] -24   weeks ago: ksh: 2022-10-18 00:00:00 != gnu: 2023-04-04 03:29:37
[BUG] -12    days ago: ksh: 2022-10-30 03:29:37 != gnu: 2022-10-30 02:29:37
[BUG] -13    days ago: ksh: 2022-10-31 03:29:37 != gnu: 2022-10-31 02:29:37
[BUG] -14    days ago: ksh: 2022-11-01 03:29:37 != gnu: 2022-11-01 02:29:37
[BUG] -15    days ago: ksh: 2022-11-02 03:29:37 != gnu: 2022-11-02 02:29:37
[BUG] -16    days ago: ksh: 2022-11-03 03:29:37 != gnu: 2022-11-03 02:29:37
[BUG] -17    days ago: ksh: 2022-11-04 03:29:37 != gnu: 2022-11-04 02:29:37
[BUG] -18    days ago: ksh: 2022-11-05 03:29:37 != gnu: 2022-11-05 02:29:37
[BUG] -19    days ago: ksh: 2022-11-06 03:29:37 != gnu: 2022-11-06 02:29:37
[BUG] -20    days ago: ksh: 2022-11-07 03:29:37 != gnu: 2022-11-07 02:29:37
[BUG] -21    days ago: ksh: 2022-11-08 03:29:37 != gnu: 2022-11-08 02:29:37
[BUG] -22    days ago: ksh: 2022-11-09 03:29:37 != gnu: 2022-11-09 02:29:37
[BUG] -23    days ago: ksh: 2022-11-10 03:29:37 != gnu: 2022-11-10 02:29:37
[BUG] -24    days ago: ksh: 2022-11-11 03:29:37 != gnu: 2022-11-11 02:29:37
[BUG]  24 minutes ago: ksh: 2022-10-18 03:05:00 != gnu: 2022-10-18 03:05:37
[BUG]  23 minutes ago: ksh: 2022-10-18 03:06:00 != gnu: 2022-10-18 03:06:37
[BUG]  22 minutes ago: ksh: 2022-10-18 03:07:00 != gnu: 2022-10-18 03:07:37
[BUG]  21 minutes ago: ksh: 2022-10-18 03:08:00 != gnu: 2022-10-18 03:08:37
[BUG]  20 minutes ago: ksh: 2022-10-18 03:09:00 != gnu: 2022-10-18 03:09:37
[BUG]  19 minutes ago: ksh: 2022-10-18 03:10:00 != gnu: 2022-10-18 03:10:37
[BUG]  18 minutes ago: ksh: 2022-10-18 03:11:00 != gnu: 2022-10-18 03:11:37
[BUG]  17 minutes ago: ksh: 2022-10-18 03:12:00 != gnu: 2022-10-18 03:12:37
[BUG]  16 minutes ago: ksh: 2022-10-18 03:13:00 != gnu: 2022-10-18 03:13:37
[BUG]  15 minutes ago: ksh: 2022-10-18 03:14:00 != gnu: 2022-10-18 03:14:37
[BUG]  14 minutes ago: ksh: 2022-10-18 03:15:00 != gnu: 2022-10-18 03:15:37
[BUG]  13 minutes ago: ksh: 2022-10-18 03:16:00 != gnu: 2022-10-18 03:16:37
[BUG]  12 minutes ago: ksh: 2022-10-18 03:17:00 != gnu: 2022-10-18 03:17:37
[BUG]  11 minutes ago: ksh: 2022-10-18 03:18:00 != gnu: 2022-10-18 03:18:37
[BUG]  10 minutes ago: ksh: 2022-10-18 03:19:00 != gnu: 2022-10-18 03:19:37
[BUG]   9 minutes ago: ksh: 2022-10-18 03:20:00 != gnu: 2022-10-18 03:20:37
[BUG]   8 minutes ago: ksh: 2022-10-18 03:21:00 != gnu: 2022-10-18 03:21:37
[BUG]   7 minutes ago: ksh: 2022-10-18 03:22:00 != gnu: 2022-10-18 03:22:37
[BUG]   6 minutes ago: ksh: 2022-10-18 03:23:00 != gnu: 2022-10-18 03:23:37
[BUG]   5 minutes ago: ksh: 2022-10-18 03:24:00 != gnu: 2022-10-18 03:24:37
[BUG]   4 minutes ago: ksh: 2022-10-18 03:25:00 != gnu: 2022-10-18 03:25:37
[BUG]   3 minutes ago: ksh: 2022-10-18 03:26:00 != gnu: 2022-10-18 03:26:37
[BUG]   2 minutes ago: ksh: 2022-10-18 03:27:00 != gnu: 2022-10-18 03:27:37
[BUG]   1 minutes ago: ksh: 2022-10-18 03:28:00 != gnu: 2022-10-18 03:28:37
[BUG]   0 minutes ago: ksh: 2022-10-18 03:29:00 != gnu: 2022-10-18 03:29:37
[BUG]  -1 minutes ago: ksh: 2022-10-18 03:30:00 != gnu: 2022-10-18 03:30:37
[BUG]  -2 minutes ago: ksh: 2022-10-18 03:31:00 != gnu: 2022-10-18 03:31:37
[BUG]  -3 minutes ago: ksh: 2022-10-18 03:32:00 != gnu: 2022-10-18 03:32:37
[BUG]  -4 minutes ago: ksh: 2022-10-18 03:33:00 != gnu: 2022-10-18 03:33:37
[BUG]  -5 minutes ago: ksh: 2022-10-18 03:34:00 != gnu: 2022-10-18 03:34:37
[BUG]  -6 minutes ago: ksh: 2022-10-18 03:35:00 != gnu: 2022-10-18 03:35:37
[BUG]  -7 minutes ago: ksh: 2022-10-18 03:36:00 != gnu: 2022-10-18 03:36:37
[BUG]  -8 minutes ago: ksh: 2022-10-18 03:37:00 != gnu: 2022-10-18 03:37:37
[BUG]  -9 minutes ago: ksh: 2022-10-18 03:38:00 != gnu: 2022-10-18 03:38:37
[BUG] -10 minutes ago: ksh: 2022-10-18 03:39:00 != gnu: 2022-10-18 03:39:37
[BUG] -11 minutes ago: ksh: 2022-10-18 03:40:00 != gnu: 2022-10-18 03:40:37
[BUG] -12 minutes ago: ksh: 2022-10-18 03:41:00 != gnu: 2022-10-18 03:41:37
[BUG] -13 minutes ago: ksh: 2022-10-18 03:42:00 != gnu: 2022-10-18 03:42:37
[BUG] -14 minutes ago: ksh: 2022-10-18 03:43:00 != gnu: 2022-10-18 03:43:37
[BUG] -15 minutes ago: ksh: 2022-10-18 03:44:00 != gnu: 2022-10-18 03:44:37
[BUG] -16 minutes ago: ksh: 2022-10-18 03:45:00 != gnu: 2022-10-18 03:45:37
[BUG] -17 minutes ago: ksh: 2022-10-18 03:46:00 != gnu: 2022-10-18 03:46:37
[BUG] -18 minutes ago: ksh: 2022-10-18 03:47:00 != gnu: 2022-10-18 03:47:37
[BUG] -19 minutes ago: ksh: 2022-10-18 03:48:00 != gnu: 2022-10-18 03:48:37
[BUG] -20 minutes ago: ksh: 2022-10-18 03:49:00 != gnu: 2022-10-18 03:49:37
[BUG] -21 minutes ago: ksh: 2022-10-18 03:50:00 != gnu: 2022-10-18 03:50:37
[BUG] -22 minutes ago: ksh: 2022-10-18 03:51:00 != gnu: 2022-10-18 03:51:37
[BUG] -23 minutes ago: ksh: 2022-10-18 03:52:00 != gnu: 2022-10-18 03:52:37
[BUG] -24 minutes ago: ksh: 2022-10-18 03:53:00 != gnu: 2022-10-18 03:53:37
```

**Comment by phidebian** [comment]
On Mon, Oct 17, 2022 at 11:05 PM Martijn Dekker ***@***.***>
wrote:

> Thanks for the PR! I'll test it now. Just a few things:
>
>    - Can you explain your changes? What does each change do and what is
>    the reason for it? This should be documented in the git commit log, so that
>    others (including myself) can understand what's going on later and possibly
>    build on your fixes. (For now you can just add a comment here, I'll make
>    sure it gets added to the commit log.)
>
> That I can do, will do in the future, will amend the non accepted PR.  Yet
I produced 'light' commit message because after a quick look at the git
log, I saw no very deep explanations of fixes, for instance I see things
like

    Fix += operator in invocation-local scopes (re: 75796a9c)

Doesn't really explain what the source fix is.

Now may be you are talking about documentation in the issue trail, where
source code and diff can be annotated, and then there each modification
could be argumented, I am no tused to that, but seems more doable may be...
so for now I will ammend my commit message unless you ask for other method.

>
>    - Please use tabs for indentation, not spaces.
>
> That I can't 'easilly', my editor, and work env, is all against TABs, tabs
are a pain to handle, both in the editor, and worst in the debugger, I am
an old school 80 cols terminal kind, I don't have an infinite number of
columns. My editor (emacs) fill gaps with SPACE, because my setup want a 2
col indentation, and will not change that.

>
>    - Please use the same coding style as the surrounding code.
>
> The coding style is broken, and then hard to follow. The rationale of
TAB's indentations as argumented by some friends programmers, is that too
deep indentation means broken code, stick with TAB and if unreadable
rethink your code. As you can see in many place in ksh code I got line soo
long that I got multi line wrap around meaning unpracticable for both
editor's and debuggers.
That beeing said, indeed I don't use the ksh coding style for my code
inspection and debugging sessions, I got to reformat/refactor the code
first, then work, debug, fix, then backport the fix the best I can in
original code, with some SPACE that can from time to time goes through.

So I try my best to provide something not 98% coding style compliant....

**Comment by phidebian** [comment]
Did you run with TZ=UTC and 'exact' in the printf, without that , your test
can't run.

The UTC thing I dunno 'yet' if this is a bug in the libast, it looks to me
that libast has hard time to applied DST traveling backward in time, and
indeed nobody can predict the future TZ table... Each years we hear people
saying this is the end of DST, yet it is still running.

The 'exact' thing is the way to avoid those :00 you got and then trig
problems.

Provided printf.sh for that purpose.

**Comment by McDutchie** [comment]
> I produced 'light' commit message because after a quick look at the git
> log, I saw no very deep explanations of fixes, for instance I see things
> like
> 
>     Fix += operator in invocation-local scopes (re: 75796a9c)
> 
> Doesn't really explain what the source fix is.

You're not looking at the complete commit messages then, you're only looking at the first line. If you view the complete commit message of that commit, you will find an explanation.

**Comment by McDutchie** [comment]
> The coding style is broken, and then hard to follow.

It's a historic coding style that in multiple ways does not follow what is currently considered best practice. It's not entirely consistent either. But I wouldn't call it broken. It does work, it's just different.

You should probably consider joining the late 20th century and abandoning the 80 column limitation.

There are also practical reasons for not changing it:

1. As comments are scarce and old RELEASE file entries are vague, we do a *lot* of code archaeology – comparing [old versions](/ksh93/ksh93-history/) to find where a particular bug got introduced. Sometimes we even have to go [back to the 1980s](/weiss/original-bsd/blob/master/local/toolchest/ksh/) to figure out why something works the way it does. As you can see, the coding style never really changed in all that time, which makes comparing very much easier.
2. Reformatting everything would make `git blame` useless for finding the commit that introduced a change of ours. 
3. Sometimes there are good reasons for the way it is, like in the main execution function, sh_exec() in xec.c. That is a huge function with deep indentation, but consists of many distinct `switch` cases for each type of command. Separating those off into their own functions would only make everything harder to follow.

**Comment by phidebian** [comment]
I don't ask for a global refactoring, nono, I just explain why <SPACE> may
comes backas <TABS>, because I do refactor the way I like it, do my work,
and de-refactor to match the coding style modulo glitches. All this is done
by tools, that the magic of it, no need to adapt to bigger display that 80
cols :), beside, though my editor/debugger are 80 cols, still I got
multiple windows side by side, even though I stay away from IDE, got my
cscope, emacs/gdb all are happy, except very long line that don't fit in
there  requiring heavy horizontal scroll :)

Actually the whole xec.c may be offuscated to fit on a single line (ala
javascript), I woud still be able to refactor (my C style) and
re-offuscate, so I have no problems to feed you back with something
approaching the current coding style :)

**Comment by phidebian** [comment]
Hi Martijn,

As you may have seen, I deleted my bug-182 branch. First of all I am not
git fluent enough to understand how I can tweak my history to avoid pollute
my branch, in short, my way of squashing my history is to pull your latest,
do my trick produce a bug-xyz branch that must be in sycn with yours (dev)
so you can merge easily... As I was rewording the commit message with
explanations, I discover that  I was commiting stuff in my br bug-182 with
some of yours in between (your pushes) and then I dunno how to squash my
stuff, may be a simple git squash do it though but didn't got time to learn
that. :)

The second reason is has I was explain my updates, it is a kinda code
review mindset then, and discover that while my fix addressed the specifics
of #182, there were portion of code that the printf.sh harness don't trig,
and I wondered if I could craft some test case for those code path (i.e not
only the %T 'ago' problem, but others), then I am working at rewriting the
printf.sh first with a more thorough harness.
I see already much more errors than just the 'ago', I will then fix them...

I am in a trip to UK now, for 3 weeks, so my spare cycles are low, yet, I
like this exercise, and will do it as time permit.

Luckily, this is a low low prio I guess, because as you said %T is
completly broken, and surely nobody cares, or else you would have plenty of
hot sites :)

Cheers,
Phi
